internal/querycoordv2/dist/dist_handler.go

```diff
@@ -181,120 +187,128 @@ func (dh *distHandler) updateSegmentsDistribution(ctx context.Context, resp *que
-			segmentInfo := dh.target.GetSealedSegment(ctx, lview.GetCollection(), ID, meta.CurrentTargetFirst)
-			if segmentInfo != nil && segmentInfo.GetLevel() == datapb.SegmentLevel_L0 {
-				continue
-			}
-			segments[ID] = &meta.Segment{
+		// Pre-allocate growing segments map
+		growings := lo.MapValues(lview.GetGrowingSegments(), func(position *msgpb.MsgPosition, id int64) *meta.Segment {
+			return &meta.Segment{
 				SegmentInfo: &datapb.SegmentInfo{
-					ID:            ID,
-					CollectionID:  lview.GetCollection(),
+					ID:            id,
+					CollectionID:  collectionID,
 					StartPosition: position,
-					InsertChannel: lview.GetChannel(),
+					InsertChannel: channelName,
 				},
-				Node: resp.NodeID,
+				Node: resp.GetNodeID(),
 			}
-		}
+		})
 
-		var version int64
-		channel, ok := channels[lview.GetChannel()]
-		if ok {
-			version = channel.GetVersion()
+		// Update DmChannel and register shard leader in same loop
+		dmChannel := &meta.DmChannel{
+			VchannelInfo: vChannelInfo,
+			Node:         resp.NodeID,
+			Version:      delegatorVersion,
+			View: &meta.LeaderView{
+				ID:                     resp.NodeID,
+				CollectionID:           collectionID,
+				Channel:                channelName,
+				Version:                delegatorVersion,
+				Segments:               lview.GetSegmentDist(),
+				GrowingSegments:        growings,
+				NumOfGrowingRows:       lview.GetNumOfGrowingRows(),
+				PartitionStatsVersions: lview.PartitionStatsVersions,
+				TargetVersion:          lview.GetTargetVersion(),
+				Status:                 lview.GetStatus(),
+			},
 		}
+		updates = append(updates, dmChannel)
 
-		view := &meta.LeaderView{
-			ID:                     resp.GetNodeID(),
-			CollectionID:           lview.GetCollection(),
-			Channel:                lview.GetChannel(),
-			Version:                version,
-			Segments:               lview.GetSegmentDist(),
-			GrowingSegments:        segments,
-			TargetVersion:          lview.TargetVersion,
-			NumOfGrowingRows:       lview.GetNumOfGrowingRows(),
-			PartitionStatsVersions: lview.PartitionStatsVersions,
+		serviceable := checkDelegatorServiceable(ctx, dh, dmChannel.View)
+		// trigger pull next target until shard leader is ready
+		if !serviceable {
+			dh.lastUpdateTs = 0
 		}
-		updates = append(updates, view)
-
-		// check leader serviceable
-		if err := utils.CheckDelegatorDataReady(dh.nodeManager, dh.target, view, meta.CurrentTarget); err != nil {
-			view.UnServiceableError = err
-			log.Ctx(ctx).
-				WithRateGroup(fmt.Sprintf("distHandler.updateLeaderView.%s", view.Channel), 1, 60).
-				RatedInfo(10, "leader is not available due to distribution not ready",
-					zap.Int64("collectionID", view.CollectionID),
-					zap.Int64("nodeID", view.ID),
-					zap.String("channel", view.Channel),
-					zap.Error(err))
-			continue
+	}
+
+	newLeaderOnNode := dh.dist.ChannelDistManager.Update(resp.GetNodeID(), updates...)
```

Comment:
Is there a guaranttee that the updates is newer than ChannelDistManager?
because I don't see anywhere compare the version